### PR TITLE
Adding dataStorage options for SolrCloud

### LINKED ISSUE #103: Support setting persistent volume reclaim policy
**Is your feature request related to a problem? Please describe.**
persistentVolumeReclaimPolicy isn't an option under PersistentVolumeClaimSpec v1 core I would like to be able to set it to "Delete" for our solrcloud test pipelines where I don't want the PVCs sticking around after deletion.

What would it take to be able set different values for persistentVolumeReclaimPolicy?

Do we need orphan pvc cleanup code logic like I see in zookeeper-operator? Or will just setting this do the cleanup for us?

**Describe the solution you'd like**
Ability to configure under dataPvcSpec "Delete" for persistentVolumeReclaimPolicy

**Describe alternatives you've considered**
In the meantime ill write a cleanup script outside of operator in our pipeline.

### PR DESCRIPTION
**Pull request by HoustonPutman**
*Issue number of the reported bug or feature request: Resolves #103*

**Describe your changes**
This PR adds a section for specifying more in-depth storage options for Solr Cloud data.

This PR also adds the ability to delete unused PVCs on deletion of the SolrCloud, or when scaling down the number of solr nodes.

**Testing performed**
Unit tests still need to be added. As well as comprehensive manual testing.

**Additional context**
Also auto-migration of old podOptions to the new location has been added.

The new storage options look like:

```golang
type SolrDataStorageOptions struct {

	// PersistentStorage is the specification for how the persistent Solr data storage should be configured.
	//
	// This option cannot be used with the "ephemeral" option.
	//
	// +optional
	PersistentStorage *SolrPersistentDataStorageOptions `json:"persistent,omitempty"`

	// EphemeralStorage is the specification for how the ephemeral Solr data storage should be configured.
	//
	// This option cannot be used with the "persistent" option.
	// Ephemeral storage is used by default if neither "persistent" or "ephemeral" is provided.
	//
	// +optional
	EphemeralStorage *SolrEphemeralDataStorageOptions `json:"ephemeral,omitempty"`

	// Options required for backups & restores to be enabled for this solrCloud.
	// +optional
	BackupRestoreOptions *SolrBackupRestoreOptions `json:"backupRestoreOptions,omitempty"`
}

type SolrPersistentDataStorageOptions struct {
	// VolumeReclaimPolicy determines how the Solr Cloud's PVCs will be treated after the cloud is deleted.
	//   - Retain: This is the default Kubernetes policy, where PVCs created for StatefulSets are not deleted when the StatefulSet is deleted.
	//   - Delete: The PVCs will be deleted by the Solr Operator after the SolrCloud object is deleted.
	// The default value is Retain, so no data will be deleted unless explicitly configured.
	// +optional
	VolumeReclaimPolicy VolumeReclaimPolicy `json:"reclaimPolicy,omitempty"`

	// PersistentVolumeClaim is the PVC object for the solr node to store its data.
	// Within metadata, the Name, Labels and Annotations are able to be specified, but defaults will be provided if necessary.
	// The entire Spec is customizable, however there will be defaults provided if necessary.
	// This field is optional. If no PVC spec is provided, then a default will be provided.
	// +optional
	PersistentVolumeClaim corev1.PersistentVolumeClaim `json:"pvc,omitempty"`
}

type VolumeReclaimPolicy string
const (
	// All pod PVCs are retained after the SolrCloud is deleted.
	VolumeReclaimPolicyRetain VolumeReclaimPolicy = "Retain"

	// All pod PVCs are deleted after the SolrCloud is deleted.
	VolumeReclaimPolicyDelete VolumeReclaimPolicy = "Delete"
)

type SolrEphemeralDataStorageOptions struct {
	//EmptyDirVolumeSource is an optional config for the emptydir volume that will store Solr data.
	// +optional
	EmptyDir corev1.EmptyDirVolumeSource `json:"emptyDir,omitempty"`
}

type SolrBackupRestoreOptions struct {
	// This is a volumeSource for a volume that will be mounted to all solrNodes to store backups and load restores.
	// The data within the volume will be namespaces for this instance, so feel free to use the same volume for multiple clouds.
	// Since the volume will be mounted to all solrNodes, it must be able to be written from multiple pods.
	// If a PVC reference is given, the PVC must have `accessModes: - ReadWriteMany`.
	// Other options are to use a NFS volume.
	Volume corev1.VolumeSource `json:"volume"`

	// Select a custom directory name to mount the backup/restore data from the given volume.
	// If not specified, then the name of the solrcloud will be used by default.
	// +optional
	Directory string `json:"directory,omitempty"`
}
```

### CODE REVIEW

**Comment by sepulworld** [comment]
A test with some checks to start...


```
--- a/controllers/solrcloud_controller_test.go
+++ b/controllers/solrcloud_controller_test.go
@@ -17,9 +17,10 @@ limitations under the License.
 package controllers

 import (
+	"testing"
+
 	zookeeperv1beta1 "github.com/pravega/zookeeper-operator/pkg/apis/zookeeper/v1beta1"
 	corev1 "k8s.io/api/core/v1"
-	"testing"

 	"github.com/bloomberg/solr-operator/controllers/util"
 	"github.com/stretchr/testify/assert"
@@ -144,6 +145,84 @@ func TestCloudReconcile(t *testing.T) {
 	expectNoIngress(g, cloudIKey)
 }

+func TestStorageVolumes(t *testing.T) {
+	SetIngressBaseUrl("")
+	UseZkCRD(true)
+	g := gomega.NewGomegaWithT(t)
+
+	instance := &solr.SolrCloud{
+		ObjectMeta: metav1.ObjectMeta{Name: expectedCloudRequest.Name, Namespace: expectedCloudRequest.Namespace},
+		Spec: solr.SolrCloudSpec{
+			ZookeeperRef: &solr.ZookeeperRef{
+				ConnectionInfo: &solr.ZookeeperConnectionInfo{
+					InternalConnectionString: "host:7271",
+				},
+			},
+			SolrJavaMem:  "-Xmx4G",
+			SolrOpts:     "extra-opts",
+			SolrLogLevel: "DEBUG",
+			StorageOptions: solr.SolrDataStorageOptions{
+				StorageType: solr.StorageTypePersistent,
+				PersistentStorage: &solr.SolrPersistentDataStorageOptions{
+					VolumeReclaimPolicy: "Delete",
+					PersistentVolumeClaim: corev1.PersistentVolumeClaim{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:   "data",
+							Labels: map[string]string{"base": "here"},
+						},
+					},
+				},
+			},
+			CustomSolrKubeOptions: solr.CustomSolrKubeOptions{
+				PodOptions: &solr.PodOptions{
+					EnvVariables:       extraVars,
+					PodSecurityContext: &podSecurityContext,
+					Volumes:            extraVolumes,
+					Affinity:           affinity,
+					Resources:          resources,
+				},
+			},
+		},
+	}
+
+	// Setup the Manager and Controller.  Wrap the Controller Reconcile function so it writes each request to a
+	// channel when it is finished.
+	mgr, err := manager.New(testCfg, manager.Options{})
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	testClient = mgr.GetClient()
+
+	solrCloudReconciler := &SolrCloudReconciler{
+		Client: testClient,
+		Log:    ctrl.Log.WithName("controllers").WithName("SolrCloud"),
+	}
+	newRec, requests := SetupTestReconcile(solrCloudReconciler)
+	g.Expect(solrCloudReconciler.SetupWithManagerAndReconciler(mgr, newRec)).NotTo(gomega.HaveOccurred())
+
+	stopMgr, mgrStopped := StartTestManager(mgr, g)
+
+	defer func() {
+		close(stopMgr)
+		mgrStopped.Wait()
+	}()
+
+	cleanupTest(g, instance.Namespace)
+
+	// Create the SolrCloud object and expect the Reconcile and StatefulSet to be created
+	err = testClient.Create(context.TODO(), instance)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer testClient.Delete(context.TODO(), instance)
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedCloudRequest)))
+
+	// Check the statefulSet
+	statefulSet := expectStatefulSet(t, g, requests, expectedCloudRequest, cloudSsKey)
+
+	assert.Equal(t, 2, len(statefulSet.Spec.Template.Spec.Volumes), "Pod has wrong number of volumes")
+	assert.Equal(t, instance.Spec.StorageOptions.PersistentStorage.PersistentVolumeClaim.Name, statefulSet.Spec.VolumeClaimTemplates[0].Name, "Data volume claim doesn't exist")
+	assert.Equal(t, "solr-cloud", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCTechnologyLabel], "Technology label doesn't match")
+	assert.Equal(t, "data", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCStorageLabel], "Storage label doesn't match")
+	assert.Equal(t, "foo-clo", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCInstanceLabel], "Instance label doesn't match")
+}
+
 func TestCustomKubeOptionsCloudReconcile(t *testing.T) {
 	ingressBaseDomain := "ing.base.domain"
 	SetIngressBaseUrl(ingressBaseDomain)
```

**Comment by sepulworld** [comment]
Was looking to see how I can assert the ReclaimPolicy is set to "Delete". Will try to figure that test out next

**Comment by sepulworld** [comment]
> Was looking to see how I can assert the ReclaimPolicy is set to "Delete". Will try to figure that test out next

Oh right, persistentVolumeReclaimPolicy is on the PV not the PVC.

**Comment by HoustonPutman** [comment]
Added your first test and it ran splendidly! I'll try to add some more to try to take some edge cases into account.

File: controllers/solrcloud_controller_storage_test.go [hunk 1]
```diff
@@ -102,6 +103,98 @@ func TestStorageVolumes(t *testing.T) {
...
+
+	// Check the statefulSet
+	statefulSet := expectStatefulSet(t, g, requests, expectedCloudRequest, cloudSsKey)
+
+	assert.Equal(t, 2, len(statefulSet.Spec.Template.Spec.Volumes), "Pod has wrong number of volumes")
+	assert.Equal(t, instance.Spec.StorageOptions.PersistentStorage.PersistentVolumeClaim.Name, statefulSet.Spec.VolumeClaimTemplates[0].Name, "Data volume claim doesn't exist")
+	assert.Equal(t, "solr-cloud", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCTechnologyLabel], "Technology label doesn't match")
+	assert.Equal(t, "data", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCStorageLabel], "Storage label doesn't match")
+	assert.Equal(t, "foo-clo", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCInstanceLabel], "Instance label doesn't match")
+	assert.Equal(t, "foo-clo", statefulSet.Spec.VolumeClaimTemplates[0].Labels[util.SolrPVCInstanceLabel], "Instance label doesn't match")
```
Comment (sepulworld) [inline controllers/solrcloud_controller_storage_test.go:121]: duplicate assert test here

**Comment by sepulworld** [comment]
@HoustonPutman we have a use case for having multiple SolrClouds use the same backupRestore path. This would help us restore data from 1 solrcloud to another without having to manipulate volume mounts and cause rolling updates.

Would you consider perhaps extending an option to override the default of the cloud name as part of the path and move to a user provided variable for this? Default to original behavior if none provided.

Current default:
https://github.com/bloomberg/solr-operator/blob/0bcff501c4e931540a2633732541837dec8463b2/controllers/util/backup_util.go#L44

Perhaps extend to offer a optional user defined path `/var/solr/solr-backup-restore/cloud/{backupDirectory}` that can then be seen my multiple SolrClouds

```
spec:
  storageOptions:
    BackupRestoreVolume:
       persistentVolumeClaim:
         claimName: my-nfs-pvc
       backupDirectory: "team-foo-solrclouds"
```

**Comment by HoustonPutman** [comment]
That's a really good use case that I completely missed. Let me cook something up.

**Comment by sepulworld** [comment]
I can post a stab at the diff

**Comment by sepulworld** [comment]
Maybe this....

```
diff --git a/api/v1beta1/solrcloud_types.go b/api/v1beta1/solrcloud_types.go
--- a/api/v1beta1/solrcloud_types.go
+++ b/api/v1beta1/solrcloud_types.go
@@ -262,6 +262,9 @@ type SolrDataStorageOptions struct {
        // Other options are to use a NFS volume.
        // +optional
        BackupRestoreVolume *corev1.VolumeSource `json:"backupRestoreVolume,omitempty"`
+
+       // Optional path for backup and restores, defaults to SolrCloud name if not provided
+       BackupRestorePath string `json:"backupRestorePath,omitempty"`
 }
```

```
diff --git a/controllers/util/solr_util.go b/controllers/util/solr_util.go
index 93879e4..ad16bab 100644
--- a/controllers/util/solr_util.go
+++ b/controllers/util/solr_util.go
@@ -109,6 +109,11 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, solrCloudStatus *solr.SolrCl
                podAnnotations = customPodOptions.Annotations
        }

+       customBackupRestorePath := solrCloud.Spec.StorageOptions.BackupRestorePath
+       if "" == customBackupRestorePath {
+               customBackupRestorePath = solrCloud.Name
+       }
+
        // Volumes & Mounts
        solrVolumes := []corev1.Volume{
                {
@@ -166,7 +171,7 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, solrCloudStatus *solr.SolrCl
                        Name:         BackupRestoreVolume,
                        VolumeSource: *solrCloud.Spec.StorageOptions.BackupRestoreVolume,
                })
-               volumeMounts = append(volumeMounts, corev1.VolumeMount{Name: BackupRestoreVolume, MountPath: BaseBackupRestorePath, SubPath: BackupRestoreSubPathForCloud(solrCloud.Name)})
+               volumeMounts = append(volumeMounts, corev1.VolumeMount{Name: BackupRestoreVolume, MountPath: BaseBackupRestorePath, SubPath: BackupRestoreSubPathForCloud(customBackupRestorePath)})
        }

        if nil != customPodOptions {
```

**Comment by HoustonPutman** [comment]
How does that look to you @sepulworld ? I will also need some help on your end testing this. I don't have very available NFS volumes to test with right now...

**Comment by sepulworld** [comment]
Nice! This should work well. We presently don't use the built in backup K8s Job feature in solr-operator, so I can't test that directly. But I can test out the custom directory option no problem.

**Comment by sepulworld** [comment]
Yes, the interface looks good to me!
```
spec:
  storageOptions:
    BackupRestoreOptions:
       volume:
         claimName: my-nfs-pvc
       directory: "team-foo-solrclouds"
```

**Comment by HoustonPutman** [comment]
I believe this PR is good to go. It has comprehensive unit tests, and I have done a lot of usage testing locally.

**Review by bsankara — APPROVED** [review]
LGTM
